### ISSUE TITLE
hint rendering is broken in qt 5.9 beta2.

### ISSUE THREAD

**Issue by amosbird** [qutebrowser/qutebrowser#2552]
As discussed in irc, here is the screenshot, 
![](https://paste.wentropy.com/CrRw.jpg)

**Comment by The-Compiler** [comment]
I wasn't able to reproduce this with the latest Qt 5.9 branch built from source. Trying the 5.9 beta 2 binary now.

**Comment by The-Compiler** [comment]
Can't reproduce it with that either... @amosbird you said you can reproduce this with `--temp-basedir`? Does `:set fonts hints 'bold 13px Monospace'` change anything? What about some other font you have installed?

**Comment by commagray** [comment]
Same, btw. Qt 5.9 beta 2 from Arch [kde-unstable] repository (but 5.8 from stable works nice). Commands above does nothing.
![screenshot_20170424_102821](https://cloud.githubusercontent.com/assets/27947961/25326450/eb02bda6-28d8-11e7-98a2-f06c0fe4176a.png)

**Comment by The-Compiler** [comment]
Does `--backend webkit` vs. `--backend webengine` (with `--temp-basedir`) make any difference?

**Comment by commagray** [comment]
Yup, on webkit everything is fine, webengine v.v. — it's sucks.

**Comment by The-Compiler** [comment]
@commagray Do you also use an IME (like fcitx) by any chance?

Can someone of you check if it happens with the IME disabled? I still can't reproduce this, but from beta 2 to beta 3 there are mostly IME changes...

(also, both QtWebKit and QtWebEngine have their good and their bad sides. That's why qutebrowser supports both.)

**Comment by commagray** [comment]
Nope, I don't. Just only X (Wayland too) and libinput, nothing else.

**Comment by The-Compiler** [comment]
Does starting with `export LIBGL_ALWAYS_SOFTWARE=1` change anything?

**Comment by commagray** [comment]
Nope, exactly the same.

**Comment by The-Compiler** [comment]
Can you reproduce it with this Python script by any chance?

```python
import sys

from PyQt5.QtCore import QUrl
from PyQt5.QtWidgets import QApplication, QLabel, QHBoxLayout, QWidget
from PyQt5.QtWebEngineWidgets import QWebEngineView


class Container(QWidget):

    def __init__(self, parent=None):
        super().__init__(parent)
        self.layout = QHBoxLayout(self)


app = QApplication(sys.argv)

container = Container()
wv = QWebEngineView()
wv.load(QUrl('https://www.qutebrowser.org/'))
container.layout.addWidget(wv)

wv.loadStarted.connect(lambda: print("Loading started"))
wv.loadProgress.connect(lambda p: print("Loading progress: {}%".format(p)))
wv.loadFinished.connect(lambda: print("Loading finished"))

label = QLabel(container)
label.setText("Hello")
label.show()
label.move(20, 20)
label.setStyleSheet("""
        color: black;
        background-color: yellow;
        font: 20pt Monospace;
""")

container.show()

app.exec_()
```

**Comment by commagray** [comment]
Nope.
![screenshot_20170424_183443](https://cloud.githubusercontent.com/assets/27947961/25345096/bfaf18ae-291c-11e7-801a-7c08269c3555.png)

**Comment by The-Compiler** [comment]
Blergh... okay, I see I really need a way to reproduce this myself - but I have no idea why it'd happen for you but not for me.

**Comment by commagray** [comment]
I think it's will be obvious and not entirely useful, but bug happening only with turned on sup-pixel rendering.

**Comment by commagray** [comment]
Stupid question: have you turned on subpixel rendering?

**Comment by The-Compiler** [comment]
Not a stupid question at all - I don't think I do! Do you have any idea what I need to do (on Arch) to reproduce this? I tried:

```
mkdir -p ~/.config/fontconfig/conf.d
ln -s /etc/fonts/conf.avail/10-sub-pixel-rgb.conf ~/.config/fontconfig/conf.d 
```

and `export FREETYPE_PROPERTIES="truetype:interpreter-version=35"` (also tried 38 and 40, all as per the [Arch wiki](https://wiki.archlinux.org/index.php/font_configuration#Subpixel_rendering))

but didn't see any change so far.

**Comment by commagray** [comment]
Welp, the only condition to reproduce a bug is
```
…
 <match target="font">
  <edit name="rgba" mode="assign">
   <const>rgb</const>
  </edit>
 </match>
…
```
in `~/.config/fontconfig/fonts.conf`, what similar to your variant with ln in `~/.config/fontconfig/conf.d `, so I have no more ideas, sorry. Lunar forces don't help at this time.

**Comment by The-Compiler** [comment]
Yay, finally reproduced with that (but not with my Terminus font) - thanks a lot! Time to bisect this to find out what Qt change exactly broke it...

**Comment by The-Compiler** [comment]
I can reproduce this with Beta 1 too - can someone of you downgrade easily and check if that's the case for you too? Unfortunately bisecting this between 5.8 and 5.9 will get quite difficult...

**Comment by The-Compiler** [comment]
Reproduced with a 5.9 qt5.git build now.

Trying with a 5.9 Qt but a 5.8 QtWebEngine next. If I still can reproduce it there, I probably need to bisect the qt5.git repo first and then bisect submodules... argh.

On the positive side, my laptop surely won't get too cold from all the compiling! 😆

**Comment by The-Compiler** [comment]
I can not reproduce it with a Qt 5.8 with QtWebEngine 5.9. Looks like an issue in qtbase then... :worried:

**Comment by The-Compiler** [comment]
```
b461645ce992b8078fe4324fb16259658e7a8cbf is the first bad commit
commit b461645ce992b8078fe4324fb16259658e7a8cbf
Author: Liang Qi <liang.qi@qt.io>
Date:   Sat Nov 26 21:11:55 2016 +0100

    Update submodules on 'dev' in qt5
    
    Change-Id: Ib6ed9acdc2de652ed0581ab2ce322bea072fec07
    Reviewed-by: Liang Qi <liang.qi@qt.io>

:160000 160000 a01f2112f4b88dbccd4af8c16248457376e363ab a0551ae6100a12a0590ef7352b5ea027a442c0d7 M	qtbase
:160000 160000 e584f8753a39d3364aa50b4b89bafa4a0209af7c 470330e69b7ee34beb27ed905fcb923d9cfdae31 M	qtdatavis3d
:160000 160000 fbfc433dae873e3cfd1247e020faff4c1303c659 af4516ac866cc877c3a033d6b837b30445ee8f1a M	qtdeclarative
:160000 160000 88edd02b2bf872af5348531d05a47d7e15a61702 b0fd258cbb2a17610a185f76076cf6cda0bdfc09 M	qtquickcontrols2
:160000 160000 fc8b7d36291611aebdc344f2e26a992da0003bbc 7561e9b64aeb649ffb3d091b1bf615efff56c730 M	qttools
:160000 160000 0e5b26942c9ff8b85322a63f577bd7a91ff0851a 74e146ac208dcbfc34c99d903e9827ef2b932046 M	qtwayland
:160000 160000 b4f8480fd41a806abe04e8eb2a4c416b2eff53c6 f2f20462e146508868d488f9bc370c53e000d2d9 M	qtwebengine
:160000 160000 fdfa61d21b09adeb1bc2e975966c7433194f3cb3 d432d42dd1cc2d5adac5fd2e7e5a55b7fbb4304d M	qtwebview
```

Still need to verify this, good chance something went wrong with so much crazy git bisecting.

**Comment by The-Compiler** [comment]
in qtbase:

```
ffd316ebe3963b7ff8d94a3484ac85de793b8299 is the first bad commit
commit ffd316ebe3963b7ff8d94a3484ac85de793b8299
Author: Allan Sandfeld Jensen <allan.jensen@qt.io>
Date:   Mon Nov 21 15:10:39 2016 +0100

    Replace QDrawHelperGammaTables with QColorProfile
    
    Turns the two set of tables in QDrawHelperGammaTables into two
    QColorProfile classes that use similar structures and can be reused for
    other gamma correction.
    
    At the same time clean-up and improve the comma-correct blending code
    to use the new profiles and QRgba64 precision.
    
    Change-Id: I302bd87a5c836e1010fff6d633eeb56fd4ae2ff0
    Reviewed-by: Olivier Goffart (Woboq GmbH) <ogoffart@woboq.com>

:040000 040000 28e8d0b7b4860b8da6c3901c1ef52879fc89aa6e 83ff819a3f374ca1e40ab0d3112d513a047ff0e8 M	src
:040000 040000 08a93e0923a2e166b8b1a8c523550262231eea04 ef8ea826c481df577dc30871f56953e4102218d8 M	tests

```

**Comment by The-Compiler** [comment]
I was able to reproduce this with a QLabel only, so this definitely doesn't seem to be related to the backend - @commagray are you sure you tested with Qt 5.9 with QtWebKit?

I reported it as [QTBUG-60469](https://bugreports.qt.io/browse/QTBUG-60469). I hope it'll be fixed for the Qt 5.9 release - as a workaround, you can set a hint background which doesn't use transparency:

```
:set colors hints.bg 'qlineargradient(x1:0, y1:0, x2:0, y2:1, stop:0 rgb(255, 247, 133), stop:1 rgb(255, 197, 66))'
```

**Comment by The-Compiler** [comment]
This is going to be fixed in Qt by [this change](https://codereview.qt-project.org/#/c/184708/) if it gets in in time, or [this patch](https://paste.kde.org/plnceffjp) if not.
